### ISSUE TITLE
Redirecting from base url on client causes page flicker

### ISSUE THREAD

**Issue by balibebas** [gatsbyjs/gatsby#25601]
## Description

Redirecting from `/` to `/foo` with `createRedirect` using the `redirectInBrowser` option set to `true` causes the page to flicker. While it's clear in the docs `createRedirect` isn't a stand-alone feature it's useful to set `redirectInBrowser` for development builds for demonstration purposes and sane dev workflow.

### Steps to reproduce

Run the [minimal reproduction](https://git.habd.as/jhabdas/gatsby-starter-default-redirect-flicker/) and navigate to the site root in a browser.

Clone default starter and remove `index.js` page:

```shell
gatsby new gatsby-default-starter && \
cd gatsby-default-starter  && \
rm src/pages/index.js
```

Add a client-side redirect from `/` to `page-2` in `gatsby-node.js` like:

```js
const siteRedirects = [
  {
    fromPath: "/",
    toPath: "/page-2",
    isPermanent: false,
    redirectInBrowser: true,
  },
]

exports.createPages = async ({ actions, graphql, reporter }) => {
  const { createRedirect } = actions
  siteRedirects.forEach(redirect => createRedirect(redirect))
}
```

Then run `gatsby develop` and access the site from the base url (e.g. http://site.example:8000).

### Expected result

Does cause flicker during client-side redirect.

### Actual result

Causes flicker showing NOT FOUND page as shown here:

![Peek 2020-07-08 17-40](https://user-images.githubusercontent.com/65664227/86909934-7356d580-c14b-11ea-9975-ea7ced3b8288.gif)

### Environment

```
System:
  OS: Linux 5.7 Arch Linux
  CPU: (8) x64 Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz
  Shell: 5.8 - /usr/bin/zsh
Binaries:
  Node: 14.5.0 - /usr/bin/node
  Yarn: 1.22.4 - /usr/bin/yarn
  npm: 6.14.6 - /usr/bin/npm
Languages:
  Python: 3.8.3 - /usr/bin/python
npmPackages:
  gatsby: ^2.23.12 => 2.23.12
  gatsby-image: ^2.4.9 => 2.4.9
  gatsby-plugin-manifest: ^2.4.14 => 2.4.14
  gatsby-plugin-offline: ^3.2.13 => 3.2.13
  gatsby-plugin-react-helmet: ^3.3.6 => 3.3.6
  gatsby-plugin-sharp: ^2.6.14 => 2.6.14
  gatsby-source-filesystem: ^2.3.14 => 2.3.14
  gatsby-transformer-sharp: ^2.5.7 => 2.5.7
npmGlobalPackages:
  gatsby: 2.22.17
```

**Comment by balibebas** [comment]
To add a little more context here. If there's a page defined at `pages/index`, that page will display momentarily instead of the of the error page and an error (shown below) logged to the console. Either way there's a page flicker.

The likely use cases for redirecting like this are:

- Launch page as I've shown above
- Maintenance page

What I ended up doing in production was to use a server-side (only) redirect to prevent users from seeing them site then, once they're authenticated calling `navigate` to send them to the base url (`/`) on the client-side. The takeaway here is that in production I'm actually relying on the fact there _isn't_ a client-side redirect. In other words, this entire issue might just be a corner case that's not worth fixing.

That said, having an streamlined way to create a launch/maintenance page like I'm doing could make a nice affordance if somehow integrated into Gatsby. Of course nice-to-haves probably aren't too high on the totem pole so I won't ban an eye-lash if this is a `wontfix`.

**Comment by balibebas** [comment]
To add a little more context here. If there's a page defined at `pages/index`, that page will display momentarily instead of the of the error page and an error-level log akin to a warning printed to the console. Either way there's a page flicker.

Two use cases for redirecting from the homepage are:

- Launch page as I've shown above
- Maintenance page

What I ended up doing in production was to use a server-side (only) redirect to prevent users from seeing the site then, once they're authenticated, calling `navigate` to send them to the base url (`/`) on the client-side. So in production I'm actually relying on the fact there _isn't_ a client-side redirect (client-side redirects being the reason I opened this in the first place). In other words, this issue might just be a corner case that's not worth fixing.

Given these consideration I'm going to close this issue and note having an streamlined way to create a launch/maintenance page like I'm doing could make a nice developer affordance.
